### PR TITLE
Bug 1891498: Show additional machine config pools while updating

### PR DESCRIPTION
**Pull request by rhamilto**
Resolves https://issues.redhat.com/browse/CONSOLE-2388

Screen Shot 2020-10-23 at 11 48 13 AM

Screen Shot 2020-10-23 at 1 14 44 PM

Screen Shot 2020-10-23 at 1 38 57 PM

Known issue:  the Nodes list page does not support dynamic role filters.  See https://issues.redhat.com/browse/CONSOLE-2454

### CODE REVIEW

File: frontend/public/locales/en/cluster-settings.json [hunk 1]
```diff
@@ -18,16 +18,13 @@
   "Channel": "Channel",
   "Channels help to control the pace of updates and recommend the appropriate release versions. Update channels are tied to a minor version of OpenShift Container Platform, for example 4.5.": "Channels help to control the pace of updates and recommend the appropriate release versions. Update channels are tied to a minor version of OpenShift Container Platform, for example 4.5.",
   "Version": "Version",
-  "Worker nodes": "Worker nodes",
-  "Worker nodes may continue to update after the update of master nodes and operators are complete.": "Worker nodes may continue to update after the update of master nodes and operators are complete.",
-  "{{updatedWorkerNodes}} of {{totalWorkerNodes}}": "{{updatedWorkerNodes}} of {{totalWorkerNodes}}",
+  "{{name}} {{resource}}": "{{name}} {{resource}}",
```
Comment (spadgett) [inline frontend/public/locales/en/cluster-settings.json:21]: I don't think we need to externalize this is everything is a parameter

File: frontend/public/components/cluster-settings/cluster-settings.tsx [hunk 2]
```diff
@@ -514,51 +515,91 @@ const UpdatesType: React.FC<UpdatesTypeProps> = ({ children }) => {
...
 };
 
+const OtherNodes: React.FC<OtherNodesProps> = ({
+  hideIfComplete,
+  machineConfigPools,
+  updateStartedTime,
+}) => {
+  const otherNodes = machineConfigPools
+    .filter((mcp) => mcp.metadata.name !== 'master' && mcp.metadata.name !== 'worker')
+    .sort((a, b) => (a.metadata.creationTimestamp > b.metadata.creationTimestamp ? 1 : -1));
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/cluster-settings.tsx:584]: This should return `0` if they are the same. I'd probably just use

```suggestion
    .sort((a, b) => a.metadata.creationTimestamp.localeCompare(b.metadata.creationTimestamp));
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare

File: frontend/public/components/cluster-settings/cluster-settings.tsx [hunk 3]
```diff
@@ -1056,19 +1075,24 @@ type UpdatesTypeProps = {
...
 };
 
 type UpdateInProgressProps = {
   desiredVersion: string;
   machineConfigPools: MachineConfigPoolKind[];
-  percentWorkerNodes: number;
-  workerMachinePoolConfig: MachineConfigPoolKind;
-  totalWorkerNodes: number;
-  updatedWorkerNodes: number;
+  workerMachineConfigPool: MachineConfigPoolKind;
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/cluster-settings.tsx:1095]: This is a cleaner API 👍

**Comment by spadgett** [comment]
We should probably convert this to a Bugzilla so we can backport to 4.6.z.

**Comment by rhamilto** [comment]
Thanks, @spadgett.  Feedback incorporated.

**Comment by rhamilto** [comment]
> We should probably convert this to a Bugzilla so we can backport to 4.6.z.

Done.

**Review by spadgett — APPROVED** [review]
/lgtm

**Comment by yapei** [comment]
Tested locally
-  on Cluster Settings page, it shows infra Nodes `0 of 1`, when clicking on `infra Nodes` it redirects user to k8s/cluster/nodes?rowFilter-node-role=infra but all nodes(master nodes, worker nodes, infra nodes) are shown there
    expected result: infra nodes filter should be applied and only infra nodes are shown 
-  Worker Nodes number seems incorrect, we can get 3 worker nodes using `oc get node --selector node-role.kubernetes.io/worker` but on console it shows `0 of 2`
    expected result: show correct number of worker nodes
```
$ oc get node --selector node-role.kubernetes.io/infra
NAME                                                       STATUS   ROLES          AGE    VERSION
qe-yapeixxx-worker-c-xxx   Ready    infra,worker   112m   v1.18.3+47c0e71
$ oc get node --selector node-role.kubernetes.io/worker
NAME                                                       STATUS   ROLES          AGE    VERSION
qe-yapeixxx-worker-a-xxx   Ready    worker         113m   v1.18.3+47c0e71
qe-yapeixxx-worker-b-xxx   Ready    worker         112m   v1.18.3+47c0e71
qe-yapeixxx-worker-c-xxx   Ready    infra,worker   112m   v1.18.3+47c0e71
$ oc get node --selector node-role.kubernetes.io/master
NAME                                                 STATUS   ROLES    AGE    VERSION
qe-yapeixxx-master-0.xxx   Ready    master   126m   v1.18.3+47c0e71
qe-yapeixxx-master-1.xxx   Ready    master   126m   v1.18.3+47c0e71
qe-yapeixxx-master-2.xxx   Ready    master   125m   v1.18.3+47c0e71
```
Screen Shot 2020-11-03 at 1 58 45 PM
Screen Shot 2020-11-03 at 2 00 27 PM

**Comment by rhamilto** [comment]
> Tested locally
> 
> * on Cluster Settings page, it shows infra Nodes `0 of 1`, when clicking on `infra Nodes` it redirects user to k8s/cluster/nodes?rowFilter-node-role=infra but all nodes(master nodes, worker nodes, infra nodes) are shown there
>   expected result: infra nodes filter should be applied and only infra nodes are shown

this is a [known issue documented above](https://github.com/openshift/console/pull/6992#issue-509102708) that will be fixed in a follow on.

> * Worker Nodes number seems incorrect, we can get 3 worker nodes using `oc get node --selector node-role.kubernetes.io/worker` but on console it shows `0 of 2`
>   expected result: show correct number of worker nodes

the custom pool nodes are worker nodes that have an additional label added to them, and [removal of the worker label is optional](https://github.com/openshift/machine-config-operator/blob/master/docs/custom-pools.md#creating-a-custom-pool), so there will be a discrepancy between the two displays if the worker node label is not removed.  in other words, this is expected behavior.

**Comment by yapei** [comment]
Thank you @rhamilto for the explanation

**Pull request merged**